### ISSUE TITLE
git lfs prune never completes after the most recent git update

### ISSUE THREAD

**Issue by samaursa** [git-lfs/git-lfs#4439]
**Describe the bug**

This has been verified by multiple colleagues of mine.

Running `git lfs prune` before git update finished normally. After the most recent git update (`2.30.2`) `git lfs prune` never finished. In fact, I am getting more objects retained, in the previous runs, before the update, I had around `108k` objects retained

![image](https://user-images.githubusercontent.com/817849/111012009-66892200-8369-11eb-9a7b-1ce709579bb5.png)

Downgrading to `2.30.1` did not help

**To Reproduce**

1. Upgrade to the latest version of git for windows (with the security fix) `v2.30.2`
2. Clone an lfs repository
3. `git lfs prune`
4. prune never completes

**Expected behavior**
`git lfs prune` should complete in a timely manner

**System environment**

```
λ git config -l
http.sslcainfo=C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
http.sslbackend=openssl
diff.astextplain.textconv=astextplain
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
credential.helper=manager-core
core.autocrlf=true
core.fscache=true
core.symlinks=false
pull.rebase=false
init.defaultbranch=master
user.name=samaursa
user.email=samaursa@tuquegames.com
user.admin=true
diff.renamelimit=10000
core.autocrlf=false
core.preloadindex=true
core.fscache=true
alias.dry=!f() { git merge-tree `git merge-base $2 $1` $2 $1; }; f
gc.auto=256
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
lfs.pruneoffsetdays=0
core.repositoryformatversion=0
core.filemode=false
core.bare=false
core.logallrefupdates=true
core.symlinks=false
core.ignorecase=true
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.master.remote=origin
branch.master.merge=refs/heads/master
```

**Output of `git lfs env`**
```
λ git lfs env
git-lfs/2.13.2 (GitHub; windows amd64; go 1.14.13; git fc664697)
git version 2.30.2.windows.1

LocalReferenceDirs=
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=0
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
AccessDownload=basic
AccessUpload=basic
DownloadTransfers=basic,lfs-standalone-file
UploadTransfers=basic,lfs-standalone-file
GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
GIT_INSTALL_ROOT=C:\Program Files\Git
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Additional context**
Our repository is very big, with the `.git/lfs` size greater than `300GB`. Usually, `git lfs prune`, with the default settings, takes a few minutes to complete. Immediately after the upgrade, the prune would never finish.

Edit: removing sensitive info

**Comment by chrisd8088** [comment]
Hey, I'm sorry you're having trouble.

Since you're using Windows, I'm not sure how much I'll be able to help as it's not a system I'm very familiar with, but as a first step, you might try running the following to get a sense of where the command stalls (on PowerShell I believe you use `$env:GIT_TRACE=1`, etc., before the command):
```
$ GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1 git lfs prune
```

You could also try:
```
$ git lfs prune --dry-run --verbose
```

I tried with a Windows VM and had no problem running a basic `git lfs prune` with a sample repo either before or after upgrading to Git for Windows 2.30.2, so it's not a universal issue, I think:
```
$ git clone https://github.com/git-lfs/lfs-test-server
$ cd lfs-test-server
$ git lfs migrate import --everything
$ git lfs prune
prune: 503 local object(s), 500 retained, done.
prune: Deleting objects: 100% (3/3), done.
$ git lfs env
git-lfs/2.13.2 (GitHub; windows amd64; go 1.14.13; git fc664697)
git version 2.30.2.windows.1
```

**Comment by samaursa** [comment]
it seems like it gets stuck with stashed files

```
20:49:44.344713 trace git-lfs: RETAIN: 5ed4e5ab3d66c1a2ffd40c8028bf3c8465140b3a04daecdea21b8c7cb6c1d8d6 stashed
20:49:44.344713 trace git-lfs: RETAIN: 855cd55899070e11f43fa7be9ffd9f7f8e291d478fdfbc7143e839fd034f201c stashed
20:49:44.344713 trace git-lfs: RETAIN: ac46990c25b4510a10fcc0dbd8aa82bcf01b51e008287aa88d09cc5e3a80576f stashed
20:49:44.344713 trace git-lfs: RETAIN: 50acbf52e72af35d094cc0881c388abeecf200d65d26fb909468a3d8316c2fb8 stashed
20:49:44.345713 trace git-lfs: RETAIN: b41081d4f6a0231e60ff4ac8740d7a195f8675448bd1d35a24bdbbcf1535941c stashed
20:49:44.345713 trace git-lfs: RETAIN: 66db2d7c6763240d557588a57524a064d31305f2f1111619237e96e4a813cb39 stashed
20:49:44.345713 trace git-lfs: RETAIN: 9fed27e6b449ac782b03d3b1af20f9d81f612296fc51b3216ca2533e2f4c3384 stashed
20:49:44.345713 trace git-lfs: RETAIN: 6487d7bb52bea0f306b469c573663b559f0a39e9a14617bb6631f1dd1ee025b4 stashed
20:49:44.345713 trace git-lfs: RETAIN: ab45aa8030b852556b0f43e0f4a85f5231c77b5d1d7d1c8992d17234f1a04062 stashed
20:49:44.345713 trace git-lfs: RETAIN: 1e158f9415431b826084f45c53b407a0ba665fb6e9ca2526b489142a3fbe01e6 stashed
20:49:44.345713 trace git-lfs: RETAIN: 29260fe59f749afcd9369246cebab39262225a1a5e783b3d5bf97a27f93ba0e6 stashed
20:49:44.345713 trace git-lfs: RETAIN: 2e841071c7fa2dafb0ef9a9002899a42845b75e879682a3ddcadaba680c99f7c stashed
20:49:44.345713 trace git-lfs: RETAIN: 4337b60a3dba84d547e60469ffdb32b1162f7fd0d5452da4d91afa6a349290fd stashed
20:49:44.345713 trace git-lfs: RETAIN: edaa8bcdd4cb82fc53932fe2cd44d9f197b592537d63d367f0cd92edfabb6644 stashed
20:49:44.345713 trace git-lfs: RETAIN: 76f418e2d5e9e2c2794d13d0a617642b1822f02e231465ef51ad83fee772d0fb stashed
20:49:44.345713 trace git-lfs: RETAIN: ad5a35d3cf3abf0b38fd3ab72281f8eeb9a34a404b76ad8012ce46a7da2abf1c stashed
20:49:44.345713 trace git-lfs: RETAIN: 728de1e1dea2307ef2a8993c0cfb10098f2e6a7f668f6b57becd2d42fe29a4c4 stashed
20:49:44.345713 trace git-lfs: RETAIN: 794d5a1b49635d1bf7e2918ec777d4bd0b9fa9e02133257cf92bb4232849d5cf stashed
20:49:44.346714 trace git-lfs: RETAIN: 6812f19ba8d64be6da4caab68869a12901b4f301791b27efc365a294bcb6acb9 stashed
20:49:44.346714 trace git-lfs: RETAIN: 7d4cbdb21dedb3fc13a86693e56b6d59379e99ba74d9238da6819500a653c0e6 stashed
20:49:44.346714 trace git-lfs: RETAIN: 482842b77646d66dd572f0653018df4d8f318d7a5f0ade45ce1444ca9ec305f7 stashed
20:49:44.346714 trace git-lfs: RETAIN: 87ee474151f4a1abb2c00fcd9bc355cde14a18b6196604671eb893f4895388b9 stashed
20:49:44.346714 trace git-lfs: RETAIN: 768d655d4bb3f2641d07d7440ca8145ba9455e7ee75c2b4e40540e07e7001cf4 stashed
20:49:44.346714 trace git-lfs: RETAIN: ac9053debf5bc3e8ef57fc9ae48d12e8de5714162004058657cb22ffa765350c stashed
20:49:44.346714 trace git-lfs: RETAIN: 324725df021933ba146c72c3378761c07eed9937daecee93335c64447891284d stashed
20:49:44.346714 trace git-lfs: RETAIN: 5a4f5fe83c0fb4607b91d65c4659d50f3702775c50e8f111c8cdc78014179a55 stashed
20:49:44.346714 trace git-lfs: RETAIN: 3dc82c06c77b9561f509be8a25a494d0bd60b15f63313523180dcd8d53908aa8 stashed
20:49:44.346714 trace git-lfs: RETAIN: 4c0fc0065dae813e38e6db1dc0b3f60bd0e0ed2247418b3e756e2a479d843455 stashed
20:49:44.346714 trace git-lfs: RETAIN: 6c7dc34b391bbe5b961621db39abd3ac0fc563678c9541583ee4f04dd13c94d8 stashed
20:49:44.346714 trace git-lfs: RETAIN: 389b701fa08ef365a485eeea8adcaae231ebf2a76cb1ac3ce45b81dae12cca4a stashed
20:49:44.346714 trace git-lfs: RETAIN: b236d917e38355973d9de8d475b361f09a6bd5b4d9071f6938b119fdcac2fea9 stashed
20:49:44.346714 trace git-lfs: RETAIN: 29a7f5cc5d738eb84f5bd0ce5fd2dbe622e918d9c7f85f1b698b9858d15141f8 stashed
20:49:44.347715 trace git-lfs: RETAIN: dc4dc612c506ffe0aef73ef5e948d51b47204a763ccdf45d7cc1cfc8780793f0 stashed
20:49:44.347715 trace git-lfs: RETAIN: 82b8a3b349ccce64690d86b9d4133ab9e08d2651a9623b5c0bd872cace1ad6ec stashed
20:49:44.347715 trace git-lfs: RETAIN: 57989d2d11478b4a2e481d84575b2c72414f7ef4841cd5ecb471917b00c3c5bf stashed
20:49:44.347715 trace git-lfs: RETAIN: af390518133d841626265b19fbfe26cf13c6c43872c484ed2e6de79a51916fdb stashed
20:49:44.347715 trace git-lfs: RETAIN: 43267d9d40a9c60a12188ca766f6d961da2146b1086ae7a1c58dd904d83b0939 stashed
20:49:44.347715 trace git-lfs: RETAIN: 39a702799f863faef83451a5d58ea95f6eef2b144d38c1aae3a68320b7b3d88e stashed
20:49:44.347715 trace git-lfs: RETAIN: e6fcdd4b81a105dd3e6e9efbb2629cea2079bef5279f4265daf62c8dd9f0ef3c stashed
20:49:44.347715 trace git-lfs: RETAIN: 527208464e2168d2f19160f98570d7abdd466c4f03e0d61760ee32847b0538a3 stashed
20:49:44.347715 trace git-lfs: RETAIN: e7ff3a6924677178192045bfa284aed305c4a4622bb1d26ca0af4d9ed889dc64 stashed
20:49:44.347715 trace git-lfs: RETAIN: f297a7a3ffce06664bdca3435e021dba83cd626d18e9a3acf1d2ddf6160c34d0 stashed
20:49:44.349716 trace git-lfs: RETAIN: 5221a1ba6acd45e0e7b82fb6b8235c592473fd7e55a37beaadf0bb55c9f42841 stashed
20:49:44.349716 trace git-lfs: RETAIN: 21cddadf3f61929e2cd340b6c52e2eeb88c84ae2b26e886144bd2520e88ef25c stashed
20:49:44.349716 trace git-lfs: RETAIN: 6b476016512ab04b83aed81d75b2054c45e1fb58429a5a16e6583925580f4197 stashed
20:49:44.349716 trace git-lfs: RETAIN: ca3efcbc04421d7f5ef375635064445c0eefff7aa97496e9ca9859be68b3e090 stashed
20:49:44.349716 trace git-lfs: RETAIN: 4cc1ad9a9bf31dfa6dff808d960880ef7144ece8e1d2deedca8854713c1a8de1 stashed
20:49:44.349716 trace git-lfs: RETAIN: 9b0ed79c5eb187fd34a9191e33a32ebceeb78048cd2070fe3a54b84ce8b1733d stashed
20:49:44.349716 trace git-lfs: RETAIN: cff6ecc07b8550a10f59baeda516ca8f104a339cfbae79e0df4758a32cef6523 stashed
20:49:44.350718 trace git-lfs: RETAIN: b1a21477c99e911f71d03d30257c7b38eda506c8cc397e257cc5d8e5d5fbf5d6 stashed
20:49:45.639888 trace git-lfs: RETAIN: e83c998f744d962714c024d59cd34e8dc57c54a7649e7623871165323a719243 stashed
20:49:45.640889 trace git-lfs: RETAIN: 16355d7ef585d01956a164f7515ac9142c01a09de6b1fcaadc0e26ad2434630b stashed
20:49:45.640889 trace git-lfs: RETAIN: 02217e14deb34218d8b27f38940e98ba57185431717abb60f4622402a25b023a stashed
20:49:45.640889 trace git-lfs: RETAIN: 763bc39b6e5bdbf82a7bcec70c00ea478accd738b73714f8ecfaeccc7dfce77e stashed
20:49:45.641890 trace git-lfs: RETAIN: 6f91e33caf62a6b3f3cb9ce36d71df4b72a4e53d7a7324a2d4d561addfad8ff7 stashed
```

**Comment by chrisd8088** [comment]
Ah, I see -- then this sounds like a duplicate of #4288, so I'm inclined to close this issue in favour of that one, but before I do, since you have a reproducible problem, can you try running `git log -g --format=%h refs/stash --`?  How long does that run?

**Comment by martinv-github** [comment]
Is this issue being investigated? Because it's definitely a legitimate one. I was also on 2.30.2 and upgrading to 2.31.0.windows.1 did not help.
Unfortunately I can only repro it on a private repo. Seems like having something stashed triggers the issue.
- I clone the repo, which has lfs objects.
- At this point, running lfs prune works just fine.
- Then I create a stash. Can be anything.
- This time around, lfs prune goes haywire with the amount of objects retained. Always ~126K ojects.
I ran it with GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1 and I get a similar output to @samaursa 

I tried reproducing with https://github.com/git-lfs/lfs-test-server but couldn't.

**Comment by chrisd8088** [comment]
As noted in https://github.com/git-lfs/git-lfs/issues/4288#issuecomment-800809950 and above, if a reproducible small test case can be created, that would be very useful.

It's also possible, but not guaranteed, that #4407 might resolve this problem, but I have a suspicion that might not be the case.  I wonder if the `git lfs prune` command is running through either a significant amount of log output at this [point](https://github.com/git-lfs/git-lfs/blob/5966ed1abed2ff6114d16e21793747071bf9ef04/lfs/gitscanner_log.go#L115), or output which includes significant volumes of error messages (in which case #4407 might help).

With your private repo that demonstrates the issue, @martinv-github, and with one or more stashed entries, could you try, first:
```
git log -g --format=%h refs/stash --
```

That should give a short list of the stash entries' SHA1s.  Then you might try something like (and I'm afraid this is a Linux version which might need to be translated to work on Windows):
```
git log -m -G 'oid sha256:' -p -U12 '--format=lfs-commit-sha: %H %P' $(git log -g --format=%h refs/stash --) > /tmp/stash.log
```

That is to say, pass the list of stash SHA1s from the first command to another `git log -m -G 'oid sha256:' -p -U12 '--format=lfs-commit-sha: %H %P'` command, and capture the standard output.  If you see error messages on your console, that might indicate the kind of overflow of stderr which #4407 tries to address.  On the other hand, if you just get a really big output file which takes an exceptionally long time to generate, that would be a different issue we'd have to address separately.

**Comment by bk2204** [comment]
I'll backport #4407 when I do a v2.13.3 release, which I'm hoping I can get done next week.  If we find that there's a different problem and we come up with a fix, I can include that as well.  But I agree with @chrisd8088 that this is a good test to see if there's a problem and see if that is the cause.  (If you want to test with the fix, there are binaries built when CI runs on `main` that you can download from the Actions tab.)

**Comment by martinv-github** [comment]
> git log -m -G 'oid sha256:' -p -U12 '--format=lfs-commit-sha: %H %P' $(git log -g --format=%h refs/stash --) > /tmp/stash.log

I got both.
The console outputted a bunch of weird messages like this
```
C:\Users\MARTIN~1.VEI\AppData\Local\Temp/p49O2j_e1.dot is not a Word Document.
```
And I also got a big output file, which, if I'm reading this correctly, is diff-ing evevy committed version of every lfs-tracked files in the repo.

**Comment by chrisd8088** [comment]
OK, thanks @martinv-github, that's helpful -- so I think there are two problems here.  The immediate one is that there's a surplus of messages from the Antiword file conversion program going to standard error when Git LFS internally requests some Git logs; that's documented in #4401 and fixed in #4407.  The "extra" error messages end up filling internal buffers and causing Git LFS to stall.

But while I haven't had time to look at it deeply yet, I'm also suspicious that we're parsing too much log output when looking for LFS blobs in the Git stash, and that that could be significantly reduced.  I'll try to find some time to look further and creating a local test case.

**Comment by martinv-github** [comment]
I went back to Git 2.29.2.windows.2 , which doesn't have the lfs issue. I ran that command again  
```git log -m -G 'oid sha256:' -p -U12 '--format=lfs-commit-sha: %H %P' $(git log -g --format=%h refs/stash --) > /tmp/stash.log```
and it yielded an identical log file.
